text/2019-02-26-tikv-client-command-line.md

```diff
@@ -0,0 +1,208 @@
+# TiKV Rust Client Command Line Tool
+
+## Summary
+
+Use the TiKV Rust Client to build a simple command line tool for inspection and anaylsis.
+
+This tool will stand as an analogue to `mycli` or `mysql-client` for users.
+
+## Motivation
+
+Currently `tikv-ctl` offers some ways to interact with a TiKV cluster, it
+focuses mostly on the *management* of a TiKV cluster instead of a *client* for
+the cluster.
+
+With the `tikv-client` crate users will find a way to interact with their TiKV
+cluster from their application. Some users will find it desirable to be able to
+inspect their cluster from a command line tool, or otherwise outside of their
+application.
+
+Building a simple command line tool to execute commands such as `get`, `set`,
+`delete`, and `scan` would mean users do not need to worry about creating such
+tools for themselves.
+
+It also provides us with a useful tool during the development cycle, as it will
+be very easy to issue one-off commands to TiKV.
+
+Further, we mentioned in the Rust Client RFC
+(https://github.com/tikv/rfcs/pull/7) we mentioned we wanted to accomplish this.
+
+## Detailed design
+
+The client will be created using the `tikv-client` crate and its already existing
+functionality. It will utilize `clap` to provide a simple wrapper around the
+crate API.
+
+### Install
+
+We will build this tool in the `tikv-cli` repository. It will be published.
+Users will be able to install it with `cargo install tikv-cli`. (See
+*Alternatives* for another option.)
+
+### Configuration
+
+The tool will retrieve its configuration from one of the following places, in
+order of precedence (greatest to least):
+
+* The command line options, such as `--pd 127.0.0.1:2379`.
+* A local configuration file `./.tikv-cli.toml`
+* The global configuration file `${XDG_CONFIG_HOME}/tikv-cli.toml`.
+* The default options (connect to localhost, no TLS).
+
+The options are:
+
+```toml
+[connection]
+# A list of `pd` endpoints.
+pd_endpoints = [
+    "127.0.0.1:2379",
+    "127.0.0.2:2379",
+]
+
+[tls]
+# The path of the CA file for TLS.
+ca_path = "./rsa.ca"
+# The path of the certificate file for TLS.
+cert_path = "./rsa.cert"
+# The path of the key file for TLS.
+key_path = "./rsa.key"
+
+[output]
+# If the ouput JSON should be minified.
+minify = false
+# Level of logging to use. Available: error. warn, info, debug, trace
+logging = "info"
+```

```

Comment:
What are the values here? I expected them to be defaults, but that doesn't match the text. (I would also expect the default log level to be `error`, or possibly `warning`.

Reply:
Should be all of these: https://docs.rs/slog/2.4.1/slog/enum.Level.html, good catch.